tests/unit/hykkt/HykktPermutationTests.hpp

```diff
@@ -1,130 +1,280 @@
 
+#include <resolve/hykkt/permutation/Permutation.hpp>
+#include <resolve/hykkt/permutation/PermutationHandler.hpp>
 #include <tests/unit/TestBase.hpp>
-#include <resolve/hykkt/Permutation.hpp>
-#include <resolve/hykkt/cpuPermutationKernels.hpp>
 
 namespace ReSolve
 {
   namespace tests
   {
     /**
      * @brief Tests for class hykkt::Permutation
-     * 
+     *
      */
     class HykktPermutationTests : public TestBase
     {
     public:
-      HykktPermutationTests() {}
-      virtual ~HykktPermutationTests() {}
+      HykktPermutationTests(memory::MemorySpace memspace = memory::HOST)
+      {
+        memspace_ = memspace;
+      }
+
+      virtual ~HykktPermutationTests()
+      {
+      }
 
-      TestOutcome permutation()
+      TestOutcome permutationTest()
       {
-        int n = 4;
-        int m = 4;
-        int nnz = 9;
-        int a_i[5]    = {0, 2, 5, 7, 9};
-        int a_j[9]    = {0, 2, 0, 1, 2, 1, 2, 1, 3};
-        int a_prc_i[5] = {0, 2, 4, 6, 9};
-        int a_prc_j[9] = {0, 3, 0, 1, 2, 3, 0, 1, 3};
-        int a_pr_i[5]  = {0, 2, 4, 6, 9};
-        int a_pr_j[9]  = {1, 2, 0, 2, 1, 3, 0, 1, 2};
-        int a_pc_i[5]  = {0, 2, 5, 7, 9};
-        int a_pc_j[9]  = {0, 1, 0, 1, 3, 0, 3, 2, 3};
-        int b_i[5] = {0}; // Initialize row pointer
-        int b_j[9] = {0}; // Initialize column indices
-        int perm[4] = {2, 0, 3, 1};
-
-        bool flagrc = false;
-        bool flagr = false;
-        bool flagc = false;
-
-        ReSolve::hykkt::Permutation pc(n, nnz, nnz);
-        pc.addHInfo(a_i, a_j);
-        pc.addJInfo(a_i, a_j, n, m);
-        pc.addJtInfo(a_i, a_j);
+        // n_hes = m_jac = 3, n_jac = 2
+        int n       = 3;
+        int m       = 2;
+        int nnz_hes = 6;
+        int nnz_jac = 4;
+
+        int  h_perm[3] = {2, 0, 1};
+        int* perm;
+        if (memspace_ == memory::HOST)
+        {
+          perm = h_perm; // already on host
+        }
+        else
+        {
+          mem_.allocateArrayOnDevice(&perm, n);
+          mem_.copyArrayHostToDevice(perm, h_perm, n);
+        }
+
+        matrix::Csr* hes    = new matrix::Csr(n, n, nnz_hes);
+        matrix::Csr* jac    = new matrix::Csr(m, n, nnz_jac);
+        matrix::Csr* jac_tr = new matrix::Csr(n, m, nnz_jac);
+        getTestData(hes, jac, jac_tr);
+
+        // correct results
+        int hes_prc_i[4] = {0, 2, 4, 6};
+        int hes_prc_j[6] = {1, 2, 0, 1, 0, 2};
+
+        int jac_pc_i[3] = {0, 2, 4};
+        int jac_pc_j[4] = {0, 1, 0, 2};
+
+        int jac_tr_pr_i[4] = {0, 2, 3, 4};
+        int jac_tr_pr_j[4] = {0, 1, 0, 1};
+
+        int result_prc_i[4] = {};
+        int result_prc_j[6] = {};
+        int result_pr_i[4]  = {};
+        int result_pr_j[4]  = {};
+        int result_pc_j[4]  = {};
+
+        ReSolve::hykkt::Permutation pc = ReSolve::hykkt::Permutation(n, m, nnz_hes, nnz_jac, memspace_);
+
+        pc.addHInfo(hes);
+        pc.addJInfo(jac);
+        pc.addJtInfo(jac_tr);
```

Comment:
```suggestion
        pc.addHInfo(&hes);
        pc.addJInfo(&jac);
        pc.addJtInfo(&jac_tr);
```
necessary change to move the matrices off the heap